### PR TITLE
IPU Integration

### PR DESCRIPTION
**Pull request by SeanNaren**
## What does this PR do?

Fixes #<issue_number>

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [ ] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pl_examples/ipu_examples/mnist.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+    def accuracy(self, logits, y):
+        # todo (sean): currently IPU poptorch doesn't implicit convert bools to tensor
+        # hence we use an explicit calculation for accuracy here. Once fixed in poptorch
+        # we can use the accuracy metric.
+        acc = torch.sum(torch.eq(torch.argmax(logits, -1), y).to(torch.float32)) / len(y)
+        return acc
+
+    def validation_epoch_end(self, outputs) -> None:
+        self.log('val_acc', torch.stack(outputs).mean(), prog_bar=True)
```
Comment (SeanNaren) [inline pl_examples/ipu_examples/mnist.py:69]: Need to clear up why this is here, not in the step itself (the step functions are jitted, and the outputs are collated from all devices, so mean averaging etc cannot happen within the functions)

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 2]
```diff
@@ -0,0 +1,239 @@
...
+        return self.process_dataloader(dataloader)
+
+    def on_reset_val_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
+
+    def on_reset_test_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
+
+    def on_reset_predict_dataloader(self, dataloader) -> Union[Iterable, DataLoader]:
+        return self.process_dataloader(dataloader)
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/ipu.py:167]: Need to be pulled into the base plugin in a separate PR if we're comfortable with adding these hooks. Basically the reason these have been introduced is `process_dataloader` assumes that the dataset size doesn't change. If the size does change, the progress bar size is messed up.

These hooks happen early enough in the code that progress bar is correct. I also looked into moving `process_dataloader` but this is finnicky.

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+        batch = apply_to_collection(batch, (torch.FloatTensor, torch.cuda.FloatTensor), function=self.batch_to)
+        return batch
+
+
+class IPUPlugin(ParallelPlugin):
+
+    def __init__(
+        self,
+        half: bool = False,
+        device_iterations: int = 1,
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/training_type/ipu.py:53]: These two options are important. `half` should be changed to a better name, but allows pure FP16 training (where the model weights are in FP16). `device_iterations` is the number of iterations to run on the device before returning (imagine running 100 batches of data on the device then returning).

it would be nice to expose them to the trainer, however this might look like overcrowding (`ipu_half` or `ipu_device_iterations`).

Because we rely on an accelerator it makes the API a bit tricky when instantiating the object:

```python
trainer = Trainer(
        accelerator=IPUAccelerator(
            training_type_plugin=IPUPlugin(device_iterations=32),
            precision_plugin=IPUPrecisionPlugin(precision=16)
        ),
        precision=16
    )
```

All that is really needed is the IPUPlugin, but I'm not sure if there is logic in place to auto-select the IPU Accelerator if we do so. then it could become:

```python
trainer = Trainer(
        plugin=IPUPlugin(device_iterations=32),
        precision=16
    )
```

File: pytorch_lightning/plugins/precision/ipu_precision.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
+from typing import Any
+
+from torch import Tensor
+
+from pytorch_lightning.plugins.precision.precision_plugin import PrecisionPlugin
+
+
+class IPUPrecisionPlugin(PrecisionPlugin):
```
Comment (SeanNaren) [inline pytorch_lightning/plugins/precision/ipu_precision.py:8]: @awaelchli said before the next boilerplate precision plugin we should refactor to have the `PrecisionPlugin` inside of the `TrainingTypePlugin`, but maybe we can allow one more if this doesn't happen in time :P

File: pytorch_lightning/plugins/training_type/ipu.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/plugins/training_type/ipu.py:137]: instead of the multiple hooks, the method `process_dataloader` could simply be the one that goes into the base class.

File: pytorch_lightning/plugins/training_type/ipu.py (hunk 3 above)
Comment (justusschock) [reply pytorch_lightning/plugins/training_type/ipu.py:53]: That logic should be easy to add I think. But that could also be part of #6090 .

One more reason to introduce something like `num_devices` and `device_type` instead of having flags for every device separately.

File: pytorch_lightning/plugins/training_type/ipu.py [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+            options = {"autoReport.all": self.autoreport}
+            if self.autoreport_dir:
+                if not os.path.exists(self.autoreport_dir):
+                    os.makedirs(self.autoreport_dir)
+                options["autoReport.directory"] = self.autoreport_dir
+            os.environ["POPLAR_ENGINE_OPTIONS"] = json.dumps(options)
+
+    def setup(self, model: Module) -> None:
+        super().setup(model)
+        if not poptorch.ipuHardwareIsAvailable():
```
Comment (justusschock) [inline pytorch_lightning/plugins/training_type/ipu.py:78]: maybe move this before the other setup? Failing as early as possible...

File: pytorch_lightning/plugins/training_type/ipu.py (hunk 2 above)
Comment (SeanNaren) [reply pytorch_lightning/plugins/training_type/ipu.py:134]: There was some confusion here; `process_dataloader` already exists in the `TrainingTypePlugin`

**Comment by SeanNaren** [comment]
will re-open
